packages/@lwc/engine/src/framework/template.ts

```diff
@@ -195,6 +200,8 @@ export function evaluateTemplate(vm: VM, html: Template): Array<VNode | null> {
                         // so we can warn if the template is attempting to use a binding
                         // that is not provided by the component instance.
                         validateFields(vm, html);
+                        // add the VM to the list of host VMs that can be re-rendered if html is swapped
+                        addHotVM(vm);
```

Comment:
after the vm and the new template were associated, the VM is added to the a set of hot VMs linked to the template, in case the template gets swapped later on.